### PR TITLE
Move ShortcutManager into Engine and separate game and engine code

### PR DESCRIPTION
**Pull request by ZehMatt**
Another piece to be moved into the Engine library, registration is now dynamic which means scripting could register shortcut actions if we had any scripting. I'm not too happy that the game asks the Config about bindings, this should be refactored later to have the Config register the bindings in ShortcutManager so we can split that dependency. Also I will eventually have the registration process done with a wrapper so we can just place the registration code next to the function, but for now its a start.

### CODE REVIEW

File: src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#include <vector>
+
+namespace OpenLoco::Input
+{
+    enum class Shortcut : uint32_t;
+}
+
+namespace OpenLoco::Input::ShortcutManager
+{
+    using string_id = uint16_t;
```
Comment (duncanspumpkin) [inline src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h:14]: this is defining string_id within the ShortcutManager namespace. I know its perfectly fine but we should stick to defining it  only ever in the OpenLoco namespace.

File: src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    struct KeyboardShortcut
```
Comment (duncanspumpkin) [inline src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h:16]: Should we come up with a new name for this struct if its no longer just for keyboards?

File: src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h (hunk 1 above)
Comment (ZehMatt) [reply src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h:14]: I'll address that, at some point string_id should be defined properly in Engine somewhere.

File: src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        Shortcut id;
+        void (*function)();
```
[the remark this answers is not in the document]
Comment (ZehMatt) [reply src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h:19]: Probably, I would ideally want to use std::function so we can use lambdas with captures but I'm also not sure if there is any interop calling this. I'll address it.

[the remark this answers is not in the document]
Comment (duncanspumpkin) [reply src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h:19]: There is no interop using this structure. Free to do what you like.

[the remark this answers is not in the document]
Comment (ZehMatt) [reply src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h:19]: Great

File: src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h (hunk 2 above)
Comment (ZehMatt) [reply src/Engine/include/OpenLoco/Engine/Input/ShortcutManager.h]: I renamed it to ShortcutEntry for the time being.

File: src/OpenLoco/src/Config.cpp [hunk 4]
```diff
@@ -111,13 +111,13 @@ namespace OpenLoco::Config
     {
         const auto& shortcutDefs = Input::ShortcutManager::getList();
         auto& shortcuts = _newConfig.shortcuts;
-        for (size_t i = 0; i < std::size(shortcuts); i++)
+        for (auto& def : shortcutDefs)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Config.cpp:114]: can be const i think

File: src/OpenLoco/src/Input/Keyboard.cpp [hunk 5]
```diff
@@ -258,18 +259,20 @@ namespace OpenLoco::Input
         auto& cfg = Config::get();
 
         // Unbind any shortcuts that may be using the current keycode.
-        for (size_t i = 0; i < ShortcutManager::kCount; i++)
+        // for (size_t i = 0; i < ShortcutManager::kCount; i++)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Input/Keyboard.cpp:262]: delete the old code

File: src/OpenLoco/src/Input/Shortcuts.cpp [hunk 6]
```diff
@@ -582,4 +468,55 @@ namespace OpenLoco::Input::ShortcutManager
     {
         GameCommands::doCommand(GameCommands::SetGameSpeedArgs{ GameSpeed::ExtraFastForward }, GameCommands::Flags::apply);
     }
+
+    void initialize()
+    {
+        ShortcutManager::add(Shortcut::closeTopmostWindow, StringIds::shortcut_close_topmost_window, closeTopmostWindow, "closeTopmostWindow", "Backspace");
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Input/Shortcuts.cpp:474]: could you play with the formatting a bit here

Comment (ZehMatt) [reply src/OpenLoco/src/Input/Shortcuts.cpp:474]: Not exactly sure what I should do, do you want them aligned or put arguments on separate lines?

Comment (duncanspumpkin) [reply src/OpenLoco/src/Input/Shortcuts.cpp:474]: Could you align things. The previous table was easier to read

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Input/Shortcuts.cpp:474]: > Could you align things. The previous table was easier to read

I'm inclined to agree, but then I authored the previous table :-)

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**